### ISSUE TITLE
Postgres backend not generating credentials

### ISSUE THREAD

**Issue by johnjelinek** [hashicorp/vault#638]
```
$ vault read postgresql/creds/readonly
Key            	Value
lease_id       	postgresql/creds/readonly/94003048-24bd-d181-e39f-15659919d860
lease_duration 	3600
lease_renewable	true
username       	token-1b6cdbad-863c-8d70-22fb-ad3e1124b9b8
password       	71e951cd-a893-291f-05cc-7e23d1a0f6d1
```

Then in postgres:
```
postgres=# \du
                                             List of roles
                 Role name                 |                   Attributes                   | Member of
-------------------------------------------+------------------------------------------------+-----------
 postgres                                  | Superuser, Create role, Create DB, Replication | {}
```

**Comment by jefferai** [comment]
Hi,

What version of Vault are you using? What policies and features do your tokens have (in particular do they have a set `num_uses`)? What SQL are you running at lease generation time? What is your postgres backend configuration?

Thanks!

**Comment by johnjelinek** [comment]
```
$ vault version
Vault v0.2.0
```

I only auth'd with the root token, ran `vault token-create` and then set that to my `$VAULT_TOKEN`. Then I mounted postgres and set up postgres connection and credential generation following the docs. My previous comment is what happens next. I'm not sure how to answer your remaining questions. I'm using a file backend configuration (I don't see postgresql as an option in the docs, only mysql, so postgres is limited to readonly credential generation).

**Comment by sethvargo** [comment]
Hey @johnjelinek 

I just tried this with Vault 0.2.0 and Vault-master (https://github.com/hashicorp/vault/commit/76b22ca24396ab720f4e5029d40872b64cb0d79b) and I am unable to reproduce the issue.

I open sourced the scripts I used here: https://github.com/sethvargo/vault-demo. In there is a setup.sh script which shows the steps for creating and mounting a postgres backend.

Please let me know if that helps diagnosing the issue at all. It would also be helpful to know what version of Postgres server you are running.

**Comment by johnjelinek** [comment]
```
$ postgres --version
postgres (PostgreSQL) 9.4.4
```

The steps in your `setup.sh` are the exact commands I ran to configure my `postgresql` mount. I know vault is connecting to the postgres server because the `vault write postgresql/config/connection` value succeeds to write. If I put in an invalid server IP/hostname, it fails.

Here's my `vault.hcl`:
```
backend "file" {
  path = "/vault"
}

listener "tcp" {
  address = "0.0.0.0:8200"
  tls_disable = 1
}
```

**Comment by johnjelinek** [comment]
This could be a helpful nugget, the client I am running `vault write postgresql/config/connection` does not have access to the database hostname/IP, but the vault server does.

**Comment by johnjelinek** [comment]
I recreated everything, this time trying the `inmem` backend, I ran through the steps in the script. Same results. I'll get you an isolated Vagrant environment to try out.

**Comment by jefferai** [comment]
What policies and features do your tokens have (in particular do they have a set num_uses)? What SQL are you running at lease generation time? What is your postgres backend configuration?

**Comment by johnjelinek** [comment]
@jefferai: I said above I don't know what any of those questions mean. Can you elaborate?

**Comment by sethvargo** [comment]
@johnjelinek I think @jefferai is asking if you have any special policies attached to _your_ token - the one you are authenticating with using the `vault` command on the client. How did you create that token? Does it have a limited number of uses? What is the lease time on that token?

What is the exact command you are running to generate the SQL tokens? Are you adding the password expiration? What do the postgres logs say on the box? What do the Vault logs say?

What is the output of `vault read postgresql/roles/readonly`?

My current working theory is that the token you are creating the configuration with is expiring. In Vault, everything is a tree and when a parent is revoked so are all the children.

**Comment by johnjelinek** [comment]
Repo: https://gist.github.com/johnjelinek/27bd2ea2a2e171fe1b04
Steps to recreate (as executed from my mac):

```
$ vagrant up --no-parallel
$ vagrant docker-logs vaultclient
$ export VAULT_ADDR=http://localhost:8200
$ export VAULT_TOKEN=50e6760c-bec8-c44d-5a12-b07693f9c034
$ vault unseal
$ vault unseal
$ vault unseal
$ vault auth

$ vault mount postgresql
Successfully mounted 'postgresql' at 'postgresql'!

$ vault write postgresql/config/connection \
  value="postgresql://postgres:postgres@db:5432/postgres?sslmode=disable"
Success! Data written to: postgresql/config/connection

$ vault write postgresql/config/lease lease=1h lease_max=24h
Success! Data written to: postgresql/config/lease

$ vault write postgresql/roles/readonly \
  sql="CREATE ROLE \"{{name}}\" WITH LOGIN PASSWORD '{{password}}' VALID UNTIL '{{expiration}}';GRANT SELECT ON ALL TABLES IN SCHEMA public TO \"{{name}}\";"
Success! Data written to: postgresql/roles/readonly

$ vault read postgresql/creds/readonly
Key            	Value
lease_id       	postgresql/creds/readonly/dbbbfd9a-2b4d-73a3-4032-04482036c84e
lease_duration 	3600
lease_renewable	true
password       	8af20881-ef28-d23f-de20-5309237b0d57
username       	root-c88e87e4-3776-e0e1-c6cb-c61090ba2f6d

$ vagrant ssh d14996d
$ docker exec -it db psql -U postgres

postgres=# \du
                             List of roles
 Role name |                   Attributes                   | Member of
-----------+------------------------------------------------+-----------
 postgres  | Superuser, Create role, Create DB, Replication | {}
```

**Comment by johnjelinek** [comment]
@sethvargo / @jefferai: I think :point_up: answers all your questions.

**Comment by jefferai** [comment]
Do you still have problems when running an official build of Vault outside of Docker?

**Comment by jefferai** [comment]
For that matter, do you still have problems when you are running *both* Postgres and Vault outside of Docker?

**Comment by johnjelinek** [comment]
I guess I could turn the Dockerfiles into scripts and execute in a vagrant environment, but it's really important for me that this works inside docker (even if I try different docker images).

**Comment by jefferai** [comment]
I think a good first step is testing the two outside of Docker. Once that's working you'll know that the setup between the two is fine and if they don't work when you move them into Docker you'll be able to troubleshoot it more easily.

I've used both Postgres and Vault inside Docker, but when using someone else's Docker image you never know what you're going to get, and that makes it pretty much impossible for us to support.

**Comment by jefferai** [comment]
Oh -- in particular, I just looked at `cgswong/vault` and it's based on Alpine. We have seen many problems with our software in Alpine, and there are plenty of other reports around of Go binaries not working correctly on Alpine, especially if they are being built inside Alpine against its libc.

If you want to run Vault in Docker I highly suggest `golang-builder` or `gockerize` -- they are far superior approaches, but because of limitations with the Docker Hub you'll never see others' images built with those tools in the public Hub.

**Comment by johnjelinek** [comment]
@jefferai: thanks! I think that may be what I was looking for. Is the limitation with alpine documented anywhere?

**Comment by jefferai** [comment]
Not that I'm aware of, because I think it's all anecdotal evidence at this point. I know, for instance, that we've seen trouble with people running Consul in Alpine-based docker images using Go 1.5.

**Comment by johnjelinek** [comment]
I created a Vagrantfile where postgres 9.4 + vault 2.0 is provisioned on boot. I can't repro the issue in that scenario, so I bet you're onto something regarding Alpine.

**Comment by johnjelinek** [comment]
I think tomorrow I'll try out your recommendations for `golang-builder` or `gockerize` and see if I make progress.

**Comment by jefferai** [comment]
Cool. Be sure to read the documentation with those about certificates (not
sure how gockerize handles it, but golang-builder has instructions).

Glad you're making progress!

**Comment by johnjelinek** [comment]
@jefferai:

I changed my Vagrantfile from https://gist.github.com/johnjelinek/27bd2ea2a2e171fe1b04 to point to an ubuntu-based vault instead of alpine and it works now:

```
  config.vm.define "vault" do |v|
    v.vm.provider "docker" do |d|
      d.name = "vault"
      d.image = "aerofs/vault"
      d.volumes = ["/var/docker/vault:/vault"]
      d.ports = ["8200:8200"]
      d.link("db:db")
      d.vagrant_vagrantfile = "./Vagrantfile.proxy"
      d.cmd = ["server", "-config=/vagrant/vault.hcl"]
      d.create_args = ["--cap-add", "IPC_LOCK", "--entrypoint", "/usr/local/bin/vault"]
    end
  end
```

Side-note: The `cgswong/vault` was using the release binary on top of alpine, not building from alpine. For `gockerize` / `golang-builder` -- would it be a valuable exercise to have vault built on top of alpine with those?

**Comment by jefferai** [comment]
> I changed my Vagrantfile from https://gist.github.com/johnjelinek/27bd2ea2a2e171fe1b04 to point to an ubuntu-based vault instead of alpine and it works now:
> Side-note: The cgswong/vault was using the release binary on top of alpine, not building from alpine. 

Yeah, so I'm not really sure how to answer this...

Vault is a static binary, so in theory it shouldn't matter whether it's running on Alpine or Ubuntu, *unless* it has some cgo stuff in there that is dynamically linking at runtime, which as far as I can tell is not the case. So this is pretty mystifying but does match up with what I've heard about Alpine/Go problems.

I think a good first step would be to try Vault with `golang-builder` and/or `gockerize` all by itself...I haven't run it myself, but it is my recommended approach because it's a totally minimal container just for running a static binary, doesn't even have a package manager or libc. So if you're up for testing, that would be useful.

**Comment by johnjelinek** [comment]
@jefferai / @sethvargo: Were you able to repro the scenario with my vagrant/docker repo above to ensure it's not just me?

**Comment by jefferai** [comment]
I honestly haven't had a chance...I'm in release mode right now and HashiConf is just around the corner (@sethvargo is already in Portland setting things up, so he's out). It's likely to be a bit before I could get to that...much apologies.

**Comment by jefferai** [comment]
Just as a point of interest, today I saw this chatter:

> on go 1.4, if you link in the net package, it needs glibc’s resolver
> this is fixed in go 1.5 as I understand it, but I haven’t had much time to play with it

Not sure if this is true, but it might be a reason for seeing failures on Alpine. Perhaps what I'd heard was backwards -- that Go stuff on Alpine was broken in 1.4 but not on 1.5, not vice-versa.

All hearsay  :-)

**Comment by johnjelinek** [comment]
So, perhaps vault 0.3.0 will not have this issue assuming it's built with Go 1.5? When does 0.3.0 ship? 



—

**Comment by jefferai** [comment]
Perhaps. And, very soon.

**Comment by johnjelinek** [comment]
Looks like alpine is having trouble with 0.3.0 too: https://github.com/cgswong/docker-vault/issues/8#issuecomment-144468104

**Comment by jefferai** [comment]
This is expected. See https://groups.google.com/forum/#!topic/vault-tool/0JCmzvk6DYA

**Comment by jefferai** [comment]
I believe this should have been fixed in 0.3.1; please re-open if not. Thanks!
